proof-language.md

```diff
@@ -0,0 +1,392 @@
+
+
+## As a stack machine
+
+It may also be viewed language interpreted using a stack machine.
+This stack machines constructs terms---either patterns or proofs. We treat
+patterns and proofs uniformly---at an abstract level, they are both terms we
+must construct, with particular well-formedness criteria. It is intended for the
+most part, that **the verifier only needs to keep proved conclusions in
+memory**.
+
+
+## Goals:
+
+-   No upfront parsing cost.
+-   Easily verifiable using ZK.
+-   Low memory usage.
+-   Composability.
+
+
+## Non-goals:
+
+-   Parallization: I have not been thinking explicitly about parallel verification of this format,
+    because I expect parallization to happen by composing units of this format into larger proofs.
+    Even so I think with some additional metadata, we could make
+    this format parallizable.
+
+-   Human readability: We aim to be a simple binary format. To become human
+    readable, we expect the user to run a secondary program to pretty-print the
+    proof. Note that symbols, lemmas, and notation do not have names.
+
+
+Terms
+=====
+
+As mentioned previously, we may view this format as a language for constructing
+terms that meet particular well-formedness criteria. Let us first enumerate
+those terms, in a python-esque pseudocode.
+
+`Term`s are of two types, `Pattern`, and `Proof`.
+Every `Term` has a well-formedness check.
+Every `Proof` has a conclusion.
+
+```python
+abstract class Term:
+    abstract def well_formed:
+        ...
+
+abstract class Pattern(Term):
+    def well_formed():
+        # Fails if any two MetaVars with the same id has
+        # different meta-requirements
+        # MetaVars are introduced later.
+    ...
+
+abstract class Proof(Term):
+    abstract def conclusion:
+        ...
+```
+
+`Pattern`s
+----------
+
+There are two classes of `Pattern`,
+the expected concrete constructors for matching logic patterns,
+and a representation for various "meta" patterns.
+
+### Standard patterns
+
+```python
+class Symbol(Pattern):
+    name: uint32
+
+class SVar(Pattern):
+    name: uint32
+
+class EVar(Pattern):
+    name: uint32
+
+class Implication(Pattern):
+    left: Pattern
+    right: Pattern
+
+class Application(Pattern):
+    left: Pattern
+    right: Pattern
+
+class Exists(Pattern):
+    var: EVar
+    subpattern : Pattern
+
+class Mu(Pattern):
+    var: SVar
+    subpattern : Pattern
+
+    def well_formed():
+        return super.well_formed()
+           and subpattern.var_occurs_positively(var)
+```
+
+### Meta-patterns
+
+Meta-patterns allow us to represent axiom- and theorem-schemas through the use
+of metavariables:
+
+```python
+class MetaVar(Pattern):
+    name: uint32
+
+    # Meta-requirements that must be satisfied by any instantiation.
+    fresh: list[EVar | SVar]        # variables that must not occur in an instatiation
+    positive: list[SVar]            # variables that must only occur positively in an instatiation
```

Comment:
Shouldn't at least `positive, negative` rather be in `Pattern`? Pattern also have positive/negative occurrences of variables, there's little reason why we should compute them. These could also be actual references in the concrete pattern.

Reply:
I'm not sure I follow. For concrete patterns, we don't need additional metadata to know whether something is positive or negative. We can just look at the term and check. For metavariables, that is not the case, since we cannot see "inside" the metavariable.

Note that a smart implementation, may keep track of positivity/negativity track in a clever way. That is not our concern here. This is not an implementation, its a specification of the language.

Reply:
Fair enough, it's a specification. I was pointing more to the current python implementation.

It is a bit redundant if you can hold direct references to them. Especially later on when we want to do substitutions more efficiently. But sure, we can do optimizations later.